profiler/src/main/java/com/splunk/opentelemetry/profiler/context/SpanContextualizer.java

```diff
@@ -0,0 +1,144 @@
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import jdk.jfr.consumer.RecordedEvent;
+import jdk.jfr.consumer.RecordedThread;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Keeps track of span scope changes and, when applicable, can wrap the RecordedEvent with span
+ * context information. This class is not thread safe.
+ */
+public class SpanContextualizer {
+
+  private static final Logger logger = LoggerFactory.getLogger(SpanContextualizer.class);
+  private final long CANT_PARSE_THREAD_ID = Long.MIN_VALUE;
+  private final Pattern lineSplitter = Pattern.compile("\n");
+  private final ThreadContextTracker threadContextTracker = new ThreadContextTracker();
+
+  /**
+   * This updates the tracked context for a This must only be called with ContextAttached events.
+   */
+  public void updateContext(RecordedEvent event) {
+    if (event.getByte("direction") == ContextAttached.IN) {
+      threadContextStarting(event);
+    } else {
+      threadContextEnding(event);
+    }
+  }
+
+  /** Parses thread info from the raw stack string and links it to a span (if available). */
+  public StackToSpanLinkage link(String rawStack) {
+    List<String> frames = Arrays.asList(lineSplitter.split(rawStack));
+    if (frames.size() < 2) {
+      // Many GC and other VM threads don't actually have a stack...
+      return StackToSpanLinkage.withoutLinkage(rawStack);
+    }
+    long threadId = parseThreadIdFromDescriptor(frames.get(0));
+    if (threadId == CANT_PARSE_THREAD_ID) {
+      return StackToSpanLinkage.withoutLinkage(rawStack);
+    }
+    return link(rawStack, threadId);
+  }
+
+  /** Links the raw stack with the span info for the given thread. */
+  StackToSpanLinkage link(String rawStack, long threadId) {
+    List<SpanLinkage> inFlightSpansForThisThread =
+        threadContextTracker.getInfLightSpansForThread(threadId);
+
+    if (inFlightSpansForThisThread.isEmpty()) {
+      // We don't know about any in-flight spans for this stack
+      return StackToSpanLinkage.withoutLinkage(rawStack);
+    }
+
+    // This thread has a span happening, augment with span details
+    if (inFlightSpansForThisThread.size() > 1) {
+      logger.debug("!! Nested spans detected: We will only use the last span for now...");
+      logger.debug(
+          "trace for thread -> {}",
+          inFlightSpansForThisThread.stream()
+              .map(SpanLinkage::getTraceId)
+              .collect(Collectors.joining(" ")));
+      logger.debug(
+          "spans for thread -> {}",
+          inFlightSpansForThisThread.stream()
+              .map(SpanLinkage::getSpanId)
+              .collect(Collectors.joining(" ")));
+    }
+    SpanLinkage spanLinkage = inFlightSpansForThisThread.get(inFlightSpansForThisThread.size() - 1);
+    return new StackToSpanLinkage(rawStack, spanLinkage);
+  }
+
+  private long parseThreadIdFromDescriptor(String descriptorLine) {
+    int secondQuote = descriptorLine.indexOf('"', 1);
```

Comment:
it might be nice to have a comment with and example of what the format looks like, just for reference when reading the code.

Reply:
That would be nice, wouldn't it..`spotlessApply` prevents me from putting a line that long.

Reply:
Also, the `ThreadDumpProcessorTest` has some examples.

Reply:
booooooo